euphonic/qpoint_frequencies.py

```diff
@@ -0,0 +1,240 @@
+                               ['frequencies_unit'])
+        super(QpointFrequencies, self).__setattr__(name, value)
+
+    def calculate_dos(self, dos_bins: Quantity) -> Spectrum1D:
+        """
+        Calculates a density of states
+
+        Parameters
+        ----------
+        dos_bins
+            Shape (n_e_bins + 1,) float Quantity. The energy bin edges
+            to use for calculating the DOS
+
+        Returns
+        -------
+        dos
+            A spectrum containing the energy bins on the x-axis and dos
+            on the y-axis
+        """
+
+        freqs = self._frequencies
+        dos_bins_unit = dos_bins.units
+        # dos_bins commonly contains a 0 bin, and converting from 0 1/cm
+        # to 0 hartree causes a RuntimeWarning, so suppress it
+        with warnings.catch_warnings():
+            warnings.filterwarnings('ignore', category=RuntimeWarning)
+            dos_bins = dos_bins.to('hartree').magnitude
+        weights = np.repeat(self.weights[:, np.newaxis],
+                            self.frequencies.shape[1],
+                            axis=1) / np.sum(self.weights)
+        dos, _ = np.histogram(freqs, dos_bins, weights=weights)
+
+        return Spectrum1D(
+            dos_bins*ureg('hartree').to(dos_bins_unit),
+            dos*ureg('dimensionless'))
+
+    def get_dispersion(self) -> Spectrum1DCollection:
+        """
+        Creates a set of 1-D bands from mode data
+
+        Bands follow the same q-point order as in the qpts array, with
+        x-axis spacing corresponding to the absolute distances between
+        q-points.  Discontinuities will appear as large jumps on the
+        x-axis.
+
+        Returns
+        -------
+        dispersion
+            A sequence of mode bands with a common x-axis
+        """
+        abscissa = _calc_abscissa(self.crystal.reciprocal_cell(), self.qpts)
+        x_tick_labels = get_qpoint_labels(self.qpts,
+                                          cell=self.crystal.to_spglib_cell())
+        return Spectrum1DCollection(abscissa, self.frequencies.T,
+                                    x_tick_labels=x_tick_labels)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Convert to a dictionary. See QpointFrequencies.from_dict for
+        details on keys/values
+        """
+        dout = _obj_to_dict(self, ['crystal', 'n_qpts', 'qpts', 'frequencies',
+                                   'weights'])
+        return dout
+
+    def to_json_file(self, filename: str) -> None:
+        """
+        Write to a JSON file. JSON fields are equivalent to
+        from_dict keys
+
+        Parameters
+        ----------
+        filename
+            Name of the JSON file to write to
+        """
+        _obj_to_json_file(self, filename)
+
+    @classmethod
+    def from_dict(cls: T, d: Dict[str, Any]) -> T:
+        """
+        Convert a dictionary to a QpointFrequencies object
+
+        Parameters
+        ----------
+        d
```

Comment:
Generally it's a good idea to avoid single-letter variable names because they aren't "greppable". It doesn't do much harm here though, as the function is only three lines long!